### PR TITLE
[RFR] Documentation being worked on

### PR DESCRIPTION
**Pull request by SimonBiggs**

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -60,11 +62,15 @@ def setup(app):
...
     'sphinxcontrib.napoleon',
-    'm2r'
+    'm2r',
+    'nbsphinx'
 ]
 
 napoleon_google_docstring = False
 
+# Remove this once decode_trf docs are live
+nbsphinx_allow_errors = True
```
Comment (SimonBiggs) [inline docs/conf.py]: @Centrus007 to ease reviewing of the new docs I have temporarily allowed docs to build even if the examples error. As you will see, without this the following example fails miserably and readthedocs fails to build:

https://pymedphys.com/en/unreleased-docs/examples/trf_mudensity.html

**Comment by SimonBiggs** [comment]
@Centrus007 this isn't ready for merging, but it certainly is ready to begin undergoing a review process. I am hoping to be able to have this merged and potentially cut a release this Friday.

To see the docs built see:

https://pymedphys.com/en/unreleased-docs/index.html

There are some examples that raise errors due to the fact that I cannot yet release the decode trf code. I do plan however to put the decode trf in this branch publicly Wednesday evening, with the aim to see if we can have it in Master by Friday.

**Comment by SimonBiggs** [comment]
Of note, there are two new documentation methods introduced. Jupyter Notebooks are being used, and live version are being served for online user testing. See https://pymedphys.com/en/unreleased-docs/examples/gamma_from_dicom.html

Also CLI interfaces are being automatically documented. See https://pymedphys.com/en/unreleased-docs/user-cli/logfile.html

**Comment by Centrus007** [comment]
This is starting to look really neat! Loving the commitment to the rework!

A few organisation notes to start with:
- The log file stuff seems to be unnecessarily separated across different sections. E.g. under "Python Library User Guide", log files have 3 relevant sections: "MU Density", "Logfile management" and "Elekta Binary logfiles". All of these probably belong in a combined section - something like Log File Utilities" or similar, with subsections. It's not clear to me that each of the above constitute separate subsections of their own, either..
- broader organisation:
   I think I like the idea of following a structure similar to the following:

```
- START HERE
    - About PyMedPhys
    - Licensing
    - FAQ
    - Release Notes (I slightly prefer this to "Changelog")
    - What's coming...

- USER GUIDE
    - Interfaces
        - Graphical (_add later_)
        - Command Line
            - Overview
            - DICOM utilities
            - Log file utilities  
    - Library

- DEVELOPER GUIDE
    - Projects
    - Recommended set-up / tools for beginners
    - Labs
    - etc.
```

**Comment by SimonBiggs** [comment]
Yup, I like it. I'll make those changes this afternoon.

**Comment by Centrus007** [comment]
Loving the CLI auto-formatting! Heading sizes could do with a little tweaking ("Positional arguments" ought to be a subheading of the CLI in question), but very handy DRY indeed!

File: docs/conf.py (hunk 1 above)
Comment (Centrus007) [reply docs/conf.py]: That did make life easier, thank you :)

**Comment by SimonBiggs** [comment]
> Loving the CLI auto-formatting! Heading sizes could do with a little tweaking ("Positional arguments" ought to be a subheading of the CLI in question), but very handy DRY indeed!

Thanks :) Leaning on the following library:

https://sphinx-argparse.readthedocs.io/en/stable/

I didn't do any of the styling myself, but we certainly could override some styles.

**Comment by Centrus007** [comment]
I'm a little less convinced, but certainly not wholly unconvinced, on the Jupyter Notebook-based docs. I quite like the Examples formatting in the API (e.g. [here](https://pymedphys.com/en/unreleased-docs/user-library/geometry.html#pymedphys.geometry.cubify_cube_definition)).

Separate, well-explained "Example" pages are a great idea, but I'm still not sure whether JNs actually add anything here. Maybe to demonstrate plots? To me, examples should match user experience, and I personally still avoid JNs because of the cumbersome need to boot a server, as well as the need to restart said server upon code changes...

**Comment by Centrus007** [comment]
Examples that illustrate code by running interactive python in Cmder honestly seems far more efficient, in my humble opinion.

**Comment by SimonBiggs** [comment]
> I'm a little less convinced, but certainly not wholly unconvinced, on the Jupyter Notebook-based docs. I quite like the Examples formatting in the API (e.g. [here](https://pymedphys.com/en/unreleased-docs/user-library/geometry.html#pymedphys.geometry.cubify_cube_definition)).
> 
> Separate, well-explained "Example" pages are a great idea, but I'm still not sure whether JNs actually add anything here. Maybe to demonstrate plots? To me, examples should match user experience, and I personally still avoid JNs because of the cumbersome need to boot a server, as well as the need to restart said server upon code changes...

One large benefit of Jupyter Notebook based examples is that without the user installing PyMedPhys they can run the notebook online. The link is provided at the top of each example page. The following is one such example:

https://mybinder.org/v2/gh/pymedphys/pymedphys/unreleased-docs?urlpath=/lab/tree/docs/examples/trf_mudensity.ipynb

I think that is a huge benefit... What do you think?

**Comment by Centrus007** [comment]
> One large benefit of Jupyter Notebook based examples is that without the user installing PyMedPhys they can run the notebook online. The link is provided at the top of each example page. The following is one such example:

Very, very good point. I missed that.

**Comment by SimonBiggs** [comment]
Heya @Centrus007,

So, I do really like your improved structure. I have implemented most of the changes you suggested. I'd be keen for feedback regarding what you think.

If you get a chance might you be able to write something (even sensible placeholder text if need be) under "About PyMedPhys" and "Roadmap"?

The overall layout does seem much nicer. I couldn't see "Projects" fitting under developer guide as I was hoping it to be a go to place if someone wants to completely implement the Logfile decoding within their own centre. I was going to write in there network configuration and other items... hmmm...

**Comment by Centrus007** [comment]
> The overall layout does seem much nicer. I couldn't see "Projects" fitting under developer guide as I was hoping it to be a go to place if someone wants to completely implement the Logfile decoding within their own centre. I was going to write in there network configuration and other items... hmmm...

Fair enough. I can also see how project contributors could legitimately be both "users" and "developers". I wasn't strong on that particular recommendation at all.

> If you get a chance might you be able to write something (even sensible placeholder text if need be) under "About PyMedPhys" and "Roadmap"?

Certainly! I'll try to contribute elsewhere as well. I also planned to go through each of the pages and review wording etc. But it might be easier to do that in a separate branch once this is merged in - working on the same branch is just asking for nightmarish conflict resolutions.

**Comment by SimonBiggs** [comment]
> Certainly! I'll try to contribute elsewhere as well. I also planned to go through each of the pages and review wording etc. But it might be easier to do that in a separate branch once this is merged in - working on the same branch is just asking for nightmarish conflict resolutions.

Could you maybe fork this branch and do a merge request into this branch? I am aiming not to merge this into master until quite late this week.

**Comment by Centrus007** [comment]
> Could you maybe fork this branch and do a merge request into this branch? I am aiming not to merge this into master until quite late this week.

Sure :)

**Comment by SimonBiggs** [comment]
@Centrus007 alright :), that's me for this evening. Time to drive home :).

File: setup.py [hunk 2]
```diff
@@ -87,21 +83,23 @@ def read(*names, **kwargs):
...
-        'xlwings >= 0.15.4',
+        'xlwings; platform_system != "Linux"',
         'dataclasses; python_version=="3.6"'
     ],
     extras_require={
         'docs': [
             'm2r',
             'sphinxcontrib-napoleon',
-            'sphinx >= 1.4',
+            'sphinx >= 1.4, < 1.8',
```
Comment (SimonBiggs) [inline setup.py]: @Centrus007, turns out a lot of the weird styling that was showing up online was due to backwards incompatibility between Sphinx 1.7 and Sphinx 1.8. I forced the build to be less than 1.8, and now the online docs look much better...

Comment (SimonBiggs) [reply setup.py]: See https://pymedphys.com/en/unreleased-docs/index.html now.

**Comment by SimonBiggs** [comment]
@Centrus007 this has the ay okay for merging. So maybe we can merge this and then any work you want to do we can do in a separate branch?

**Comment by SimonBiggs** [comment]
This document works on the docs quite a bit. I'd appreciate a few reviewers if possible. See https://pymedphys.com/en/unreleased-docs/index.html for the docs built from this branch.

File: tests/test_trfdecode/test_decode.py [hunk 3]
```diff
@@ -44,11 +42,6 @@
     os.path.dirname(__file__), "../data/trf")
 
 
-pytest.skip(
-    "Skipping decode_trf tests, the library can't be made public.",
-    allow_module_level=True)
-
```
Comment (SimonBiggs) [inline tests/test_trfdecode/test_decode.py]: @Centrus007 :) Ahh, it be nice to be running these tests once more :)

**Comment by SimonBiggs** [comment]
Heya @Centrus007,

I've merged in the changes you made and removed the FAQ that needs more thought. I'm keen for this to be merged in as is. We can work more on it in another branch?

File: tests/test_trfdecode/test_decode.py (hunk 3 above)
Comment (Centrus007) [reply tests/test_trfdecode/test_decode.py]: Woot!

**Comment by SimonBiggs** [comment]
Ahh, now that is a celebratory merge :)

**Comment by kingrpaul** [comment]
Oops. Looks like I am late to respond. Sorry about that. Look forward to seeing the outcome. 
- Paul

**Comment by SimonBiggs** [comment]
All good, still worth having a look through. If you see any issues it'd be nice to know if they're there :).

A release was cut yesterday, so that means these docs are now in stable. If there is an issue now's a good time to find it because if need be I can still retag 0.7.2 within GitHub as nothing has yet changed in the package itself.

**Comment by kingrpaul** [comment]
G:\Python\pymedphys\pymedphys>sphinx-autobuild docs docs/_build/html
Extension error: Could not import extension sphinxarg.ext
pip install sphinxarg   ?->nope

**Comment by SimonBiggs** [comment]
Run:

pip install -e .[docs]

That will install all requirements for the documentation.

**Comment by kingrpaul** [comment]
> pip install -e .[docs]

Thanks. Will do. It does look like you have been busy.

**Comment by SimonBiggs** [comment]
Yeah...

File: setup.py (hunk 2 above)
Comment (Centrus007) [reply setup.py]: @SimonBiggs, this was actually a great pick up, mate! I think it was this issue that was causing some less than ideal docstring formatting. Sphinx's parameter auto-documentation is looking a lot better, especially when the parameter type is supplied next to the parameter name. E.g. see [the gamma_shell() docstring in 0.7.x](https://pymedphys.com/en/stable/user/library/gamma.html#pymedphys.gamma.gamma_shell) to the 0.6.0 version.

**Comment by SimonBiggs** [comment]
Ahh true. Good to hear. Yeah, that was a bit of a rabbit hole trying to
solve which dependency made the docs look less than ideal. Glad it's
improved a few things :).

**Pull request merged**